# Wheat Yield Data

Ella: 
Start cleaning, beging initial preprocessing
- standardising names
- filter for wheat crop
- remove duplicates
- handle missing values

## Data Loading

In [1]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Load dataset 
df = pd.read_csv('Production_Crops_Livestock_E_All_Data.csv')

C:\Users\mason\AppData\Local\Temp\ipykernel_9588\4035631550.py:2: DtypeWarning: Columns (11,14,17,20,23,26,29,32,35,38,41,44,47,50,53,56,59,62,65,68,71,74,77,80,83,86,89,92,95,98,101,104,107,110,113,116,119,122,125,128,131,134,137,140,143,146,149,152,155,158,161,164,167,170,173,176,179,182,185,188,191,194,197) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Production_Crops_Livestock_E_All_Data.csv')


#### Initial Date Inspection

In [3]:
# Display first few rows
df.head()

Area Code Area Code (M49)         Area  Item Code Item Code (CPC)  \
0          2            '004  Afghanistan        221          '01371   
1          2            '004  Afghanistan        221          '01371   
2          2            '004  Afghanistan        221          '01371   
3          2            '004  Afghanistan        711          '01654   
4          2            '004  Afghanistan        711          '01654   

                                                Item  Element Code  \
0                                  Almonds, in shell          5312   
1                                  Almonds, in shell          5412   
2                                  Almonds, in shell          5510   
3  Anise, badian, coriander, cumin, caraway, fenn...          5312   
4  Anise, badian, coriander, cumin, caraway, fenn...          5412   

          Element   Unit  Y1961  ... Y2020N    Y2021  Y2021F Y2021N    Y2022  \
0  Area harvested     ha    0.0  ...    NaN  36862.0       A    NaN  36462.0   
1           Yield  kg/ha    NaN  ...    NaN   1743.2       A    NaN   1742.0   
2      Production      t    0.0  ...    NaN  64256.0       A    NaN  63515.0   
3  Area harvested     ha    NaN  ...    NaN  25357.0       E    NaN  25403.0   
4           Yield  kg/ha    NaN  ...    NaN    705.0       E    NaN    704.0   

   Y2022F Y2022N    Y2023  Y2023F Y2023N  
0       A    NaN  37000.0       A    NaN  
1       A    NaN   1810.8       A    NaN  
2       A    NaN  67000.0       A    NaN  
3       E    NaN  25439.0       E    NaN  
4       E    NaN    704.0       E    NaN  

[5 rows x 198 columns]

In [4]:
# Check dataset structure and data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79711 entries, 0 to 79710
Columns: 198 entries, Area Code to Y2023N
dtypes: float64(63), int64(3), object(132)
memory usage: 120.4+ MB


In [5]:
# View summary statistics
df.describe()

Area Code     Item Code  Element Code         Y1961         Y1962  \
count  79711.000000  79711.000000  79711.000000  5.737000e+04  5.737700e+04   
mean    1313.302568    865.879251   5408.833310  1.514965e+06  1.549240e+06   
std     2194.895639   1461.515725    101.505636  1.586835e+07  1.615213e+07   
min        1.000000     15.000000   5111.000000  0.000000e+00  0.000000e+00   
25%       89.000000    358.000000   5312.000000  8.362500e+02  8.625000e+02   
50%      169.000000    587.000000   5412.000000  7.500000e+03  7.700000e+03   
75%      272.000000   1037.000000   5510.000000  6.655310e+04  6.874600e+04   
max     5817.000000  17530.000000   5513.000000  1.342210e+09  1.360092e+09   

              Y1963         Y1964         Y1965         Y1966         Y1967  \
count  5.737600e+04  5.739100e+04  5.738600e+04  5.744700e+04  5.745100e+04   
mean   1.582932e+06  1.630178e+06  1.663916e+06  1.704733e+06  1.749159e+06   
std    1.645200e+07  1.694783e+07  1.731071e+07  1.775626e+07  1.817071e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    8.929750e+02  9.100000e+02  9.343750e+02  9.500000e+02  9.885000e+02   
50%    7.981640e+03  8.000000e+03  8.130950e+03  8.400000e+03  8.647000e+03   
75%    7.060525e+04  7.305150e+04  7.500000e+04  7.721212e+04  7.975700e+04   
max    1.368767e+09  1.381032e+09  1.397533e+09  1.408673e+09  1.428836e+09   

       ...         Y2014         Y2015         Y2016         Y2017  \
count  ...  7.103000e+04  7.109600e+04  7.118300e+04  7.141000e+04   
mean   ...  3.156620e+06  3.184441e+06  3.226641e+06  3.273442e+06   
std    ...  3.576517e+07  3.611638e+07  3.656186e+07  3.692842e+07   
min    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    ...  1.524205e+03  1.543307e+03  1.560000e+03  1.557030e+03   
50%    ...  1.207270e+04  1.223683e+04  1.234740e+04  1.237300e+04   
75%    ...  1.211520e+05  1.225511e+05  1.234405e+05  1.241075e+05   
max    ...  2.818170e+09  2.837867e+09  2.912244e+09  2.961412e+09   

              Y2018         Y2019         Y2020         Y2021         Y2022  \
count  7.072900e+04  7.082200e+04  7.093500e+04  7.099100e+04  7.112000e+04   
mean   3.329584e+06  3.345364e+06  3.384600e+06  3.464475e+06  3.504832e+06   
std    3.735102e+07  3.722087e+07  3.759316e+07  3.870242e+07  3.929182e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    1.442300e+03  1.433050e+03  1.432300e+03  1.437730e+03  1.441750e+03   
50%    1.222000e+04  1.225000e+04  1.218570e+04  1.214760e+04  1.220000e+04   
75%    1.255072e+05  1.269745e+05  1.263125e+05  1.270417e+05  1.278680e+05   
max    2.901857e+09  2.964207e+09  3.008965e+09  3.070651e+09  3.073689e+09   

              Y2023  
count  6.614500e+04  
mean   3.752891e+06  
std    4.156594e+07  
min    0.000000e+00  
25%    1.389000e+03  
50%    1.174500e+04  
75%    1.227800e+05  
max    3.134294e+09  

[8 rows x 66 columns]

Observations
- 79711 rows, 198 colu,ns 
- Data types floats for numeric columns (yields, area, production), object for strings and meta data
- Some columns have N, F this could be forecasted data

Missing values 
- Some years have NAN for certain crops in some countries 
- This will need to be handled in data cleaning

Range and Scale
- Yield is in hundreds to thousands
- Area harvested and prduction can go up to billions
- There are outliers

## Data Cleaning

In [6]:
# Standardise column names
# Start with country names
# Remove quotes and extra spaces for consistency
df['Area'] = df['Area'].str.replace("'", "").str.strip()

# Clean crop names by removing extra spaces 
df['Item'] = df['Item'].str.strip()

df[['Area', 'Item']].head()

Area                                               Item
0  Afghanistan                                  Almonds, in shell
1  Afghanistan                                  Almonds, in shell
2  Afghanistan                                  Almonds, in shell
3  Afghanistan  Anise, badian, coriander, cumin, caraway, fenn...
4  Afghanistan  Anise, badian, coriander, cumin, caraway, fenn...

In [7]:
# Filter to Wheat data only

# Create a dataframe for wheat production only
wheat_df = df[df['Item'] == 'Wheat']

# Check first few rows
wheat_df.head()

Area Code Area Code (M49)         Area  Item Code Item Code (CPC)   Item  \
196          2            '004  Afghanistan         15           '0111  Wheat   
197          2            '004  Afghanistan         15           '0111  Wheat   
198          2            '004  Afghanistan         15           '0111  Wheat   
539          3            '008      Albania         15           '0111  Wheat   
540          3            '008      Albania         15           '0111  Wheat   

     Element Code         Element   Unit      Y1961  ... Y2020N      Y2021  \
196          5312  Area harvested     ha  2230000.0  ...    NaN  2046564.0   
197          5412           Yield  kg/ha     1022.0  ...    NaN     1963.1   
198          5510      Production      t  2279000.0  ...    NaN  4017657.0   
539          5312  Area harvested     ha   126500.0  ...    NaN    54514.0   
540          5412           Yield  kg/ha      773.1  ...    NaN     4130.5   

     Y2021F Y2021N      Y2022  Y2022F Y2022N      Y2023  Y2023F  \
196       A    NaN  1859339.0       A    NaN  2003070.0       I   
197       A    NaN     2045.3       A    NaN     2146.7       E   
198       A    NaN  3802895.0       A    NaN  4300000.0       X   
539       A    NaN    54267.0       A    NaN    57620.0       A   
540       A    NaN     4296.3       A    NaN     4053.9       A   

                Y2023N  
196                NaN  
197                NaN  
198  Unofficial figure  
539                NaN  
540                NaN  

[5 rows x 198 columns]

In [8]:
# Filter for rows where the element is yield 
wheat_df = wheat_df[wheat_df['Element'] == 'Yield']

wheat_df.head

<bound method NDFrame.head of        Area Code Area Code (M49)                                     Area  \
197            2            '004                              Afghanistan   
540            3            '008                                  Albania   
889            4            '012                                  Algeria   
1125           7            '024                                   Angola   
1720           9            '032                                Argentina   
...          ...             ...                                      ...   
77152       5801            '199                Least Developed Countries   
77806       5802            '432         Land Locked Developing Countries   
78299       5803            '722           Small Island Developing States   
78958       5815            '901        Low Income Food Deficit Countries   
79640       5817            '902  Net Food Importing Developing Countries   

       Item Code Item Code (CPC)   Item  Elem

In [9]:
# Investigate column structure for columns
wheat_df.columns


Index(['Area Code', 'Area Code (M49)', 'Area', 'Item Code', 'Item Code (CPC)',
       'Item', 'Element Code', 'Element', 'Unit', 'Y1961',
       ...
       'Y2020N', 'Y2021', 'Y2021F', 'Y2021N', 'Y2022', 'Y2022F', 'Y2022N',
       'Y2023', 'Y2023F', 'Y2023N'],
      dtype='object', length=198)

In [10]:
# Count year-related columns
# This includes data columns, flags (F), and notes (N) for years 1961-2023")
year_columns = [col for col in df.columns if col.startswith("Y")]
print(f"Total year-related columns: {len(year_columns)}")

Total year-related columns: 189


In [11]:
# Missing values 
# Check missing values
print("Missing values per column:")
wheat_df.isnull().sum()

Missing values per column:


Area Code            0
Area Code (M49)      0
Area                 0
Item Code            0
Item Code (CPC)      0
                  ... 
Y2022F              10
Y2022N             164
Y2023               10
Y2023F              10
Y2023N             164
Length: 198, dtype: int64

In [12]:
# Missing values 
# Check missing values
print("Missing values per column:")
wheat_df.isnull().sum()

Missing values per column:


Area Code            0
Area Code (M49)      0
Area                 0
Item Code            0
Item Code (CPC)      0
                  ... 
Y2022F              10
Y2022N             164
Y2023               10
Y2023F              10
Y2023N             164
Length: 198, dtype: int64

In [13]:
# Drop duplicate rows if any 
wheat_df = wheat_df.drop_duplicates()
wheat_df.shape

(164, 198)

**Note for Mason:**

From this point onwards, please handle:
1. Removing flag and note columns (Y****F and Y****N)
2. Removing rows with completely missing data
3. Any other missing value treatment needed for analysis

The current dataset `wheat_df` contains:
- 164 rows (countries with wheat yield data)
- 198 columns (including year data, flags, and notes)
- Element filtered to 'Yield' only


In [14]:
#remove flag and note columns
wheat_df = wheat_df.drop(columns=[col for col in wheat_df.columns if col.endswith('F') or col.endswith('N')])
wheat_df.shape

(164, 72)

In [15]:
#redo year_columns after dropping flag and note columns
year_columns = [col for col in wheat_df.columns if col.startswith("Y")]
#drop rows where all year data is missing
wheat_df = wheat_df.dropna(how='all', subset=year_columns)
wheat_df

Area Code Area Code (M49)                                     Area  \
197            2            '004                              Afghanistan   
540            3            '008                                  Albania   
889            4            '012                                  Algeria   
1125           7            '024                                   Angola   
1720           9            '032                                Argentina   
...          ...             ...                                      ...   
77152       5801            '199                Least Developed Countries   
77806       5802            '432         Land Locked Developing Countries   
78299       5803            '722           Small Island Developing States   
78958       5815            '901        Low Income Food Deficit Countries   
79640       5817            '902  Net Food Importing Developing Countries   

       Item Code Item Code (CPC)   Item  Element Code Element   Unit   Y1961  \
197           15           '0111  Wheat          5412   Yield  kg/ha  1022.0   
540           15           '0111  Wheat          5412   Yield  kg/ha   773.1   
889           15           '0111  Wheat          5412   Yield  kg/ha   406.0   
1125          15           '0111  Wheat          5412   Yield  kg/ha  1111.1   
1720          15           '0111  Wheat          5412   Yield  kg/ha  1295.0   
...          ...             ...    ...           ...     ...    ...     ...   
77152         15           '0111  Wheat          5412   Yield  kg/ha   935.1   
77806         15           '0111  Wheat          5412   Yield  kg/ha   948.9   
78299         15           '0111  Wheat          5412   Yield  kg/ha  2500.0   
78958         15           '0111  Wheat          5412   Yield  kg/ha   835.3   
79640         15           '0111  Wheat          5412   Yield  kg/ha   853.9   

       ...   Y2014   Y2015   Y2016   Y2017   Y2018   Y2019   Y2020   Y2021  \
197    ...  2023.7  2195.9  1980.3  2034.2  2210.0  2095.1  1943.4  1963.1   
540    ...  4000.1  3951.1  3900.0  4036.7  3692.7  4068.0  4327.1  4130.5   
889    ...  1475.3  1464.0  1183.3  1150.1  2043.3  1963.0  1681.1  1584.3   
1125   ...   977.9   962.2   946.6   931.0   915.3   879.8   652.0   931.0   
1720   ...  2661.9  2810.0  2862.3  3304.7  3180.6  3216.0  2938.7  2759.5   
...    ...     ...     ...     ...     ...     ...     ...     ...     ...   
77152  ...  2291.8  2520.8  2304.2  2424.5  2524.9  2523.2  2473.3  2577.7   
77806  ...  1770.8  1888.8  1874.3  1861.9  1853.8  1851.4  1889.1  1748.6   
78299  ...     NaN   702.2   654.3   601.7   522.4   592.8   509.4   561.8   
78958  ...  2525.6  2851.4  2581.9  2579.8  2575.3  2794.7  2725.7  2617.8   
79640  ...  2677.6  2807.0  2644.4  2820.2  2890.9  2765.3  2653.1  3032.3   

        Y2022   Y2023  
197    2045.3  2146.7  
540    4296.3  4053.9  
889    1666.7  1388.9  
1125    931.0   931.0  
1720   3382.5  2287.6  
...       ...     ...  
77152  2588.3  2615.3  
77806  1922.4  1777.3  
78299   550.3   537.7  
78958  2754.1  2921.3  
79640  2811.9  3007.7  

[164 rows x 72 columns]

In [16]:
#remove apostraphes from columns
wheat_df['Area Code (M49)'] = wheat_df['Area Code (M49)'].str.replace("'" , "")
wheat_df['Item Code (CPC)'] = wheat_df['Item Code (CPC)'].str.replace("'" , "")
wheat_df

Area Code Area Code (M49)                                     Area  \
197            2             004                              Afghanistan   
540            3             008                                  Albania   
889            4             012                                  Algeria   
1125           7             024                                   Angola   
1720           9             032                                Argentina   
...          ...             ...                                      ...   
77152       5801             199                Least Developed Countries   
77806       5802             432         Land Locked Developing Countries   
78299       5803             722           Small Island Developing States   
78958       5815             901        Low Income Food Deficit Countries   
79640       5817             902  Net Food Importing Developing Countries   

       Item Code Item Code (CPC)   Item  Element Code Element   Unit   Y1961  \
197           15            0111  Wheat          5412   Yield  kg/ha  1022.0   
540           15            0111  Wheat          5412   Yield  kg/ha   773.1   
889           15            0111  Wheat          5412   Yield  kg/ha   406.0   
1125          15            0111  Wheat          5412   Yield  kg/ha  1111.1   
1720          15            0111  Wheat          5412   Yield  kg/ha  1295.0   
...          ...             ...    ...           ...     ...    ...     ...   
77152         15            0111  Wheat          5412   Yield  kg/ha   935.1   
77806         15            0111  Wheat          5412   Yield  kg/ha   948.9   
78299         15            0111  Wheat          5412   Yield  kg/ha  2500.0   
78958         15            0111  Wheat          5412   Yield  kg/ha   835.3   
79640         15            0111  Wheat          5412   Yield  kg/ha   853.9   

       ...   Y2014   Y2015   Y2016   Y2017   Y2018   Y2019   Y2020   Y2021  \
197    ...  2023.7  2195.9  1980.3  2034.2  2210.0  2095.1  1943.4  1963.1   
540    ...  4000.1  3951.1  3900.0  4036.7  3692.7  4068.0  4327.1  4130.5   
889    ...  1475.3  1464.0  1183.3  1150.1  2043.3  1963.0  1681.1  1584.3   
1125   ...   977.9   962.2   946.6   931.0   915.3   879.8   652.0   931.0   
1720   ...  2661.9  2810.0  2862.3  3304.7  3180.6  3216.0  2938.7  2759.5   
...    ...     ...     ...     ...     ...     ...     ...     ...     ...   
77152  ...  2291.8  2520.8  2304.2  2424.5  2524.9  2523.2  2473.3  2577.7   
77806  ...  1770.8  1888.8  1874.3  1861.9  1853.8  1851.4  1889.1  1748.6   
78299  ...     NaN   702.2   654.3   601.7   522.4   592.8   509.4   561.8   
78958  ...  2525.6  2851.4  2581.9  2579.8  2575.3  2794.7  2725.7  2617.8   
79640  ...  2677.6  2807.0  2644.4  2820.2  2890.9  2765.3  2653.1  3032.3   

        Y2022   Y2023  
197    2045.3  2146.7  
540    4296.3  4053.9  
889    1666.7  1388.9  
1125    931.0   931.0  
1720   3382.5  2287.6  
...       ...     ...  
77152  2588.3  2615.3  
77806  1922.4  1777.3  
78299   550.3   537.7  
78958  2754.1  2921.3  
79640  2811.9  3007.7  

[164 rows x 72 columns]

This looks clean to me bar the NaN values. We just need to work out collectively what we want to do with them. I am not sure that getting rid of any row with them in is viable. Last edit - 1st december